### PR TITLE
custom locator strategy

### PR DESCRIPTION
**Pull request by baruchvlz**
## Proposed changes
Related issue: #4578

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mgrybyk** [comment]
How should it work in the end, can you please provide with examples?

**Comment by baruchvlz** [comment]
@mgrybyk 

@christian-bromann and I had a conversation about how to approach this. Essentially this is what we came up with.

The user will have to register a locator somewhere in their tests like so

```ts
browser.addLocatorStrategy(name: string, callback: (args: any[]) => HTMLElement))
```

When they want to use this strategy, they can call it as such:

```ts
browser.custom$(name: string, strategyArgs: any[]) => WDIOElement
```

Any suggestions / concerns let me know.

**Comment by CrispusDH** [comment]
@baruchvlz , I think might be better to create two methods:
```js
browser.custom$(name: string, strategyArgs: any[]) => WDIOElement
browser.custom$$(name: string, strategyArgs: any[]) => WDIOElement[]
```
in such case it would be convenient to work with types and be predictable what you expect to get.
What do you think?

**Comment by baruchvlz** [comment]
@CrispusDH Yea that makes sense, I'll add the commands.

**Comment by baruchvlz** [comment]
@CrispusDH 
Sorry, on second thought, having both `custom$` and `custom$$` might be confusing since you have to add the strategies before you use them. It would be better to consider renaming the command to some more explicit like`useLocatorStrategy`, this way its easier to understand while writing tests, and we always return one element since it depends on the script passed by the user.

Alternatively, we could add an extra flag to the command for selecting multiple elements.

```ts
useLocatorStrategy<T = {}>(name: string, strategyArgument: string | T, multiple: boolean) => WDIOElement | WDIOElement[]
```

Adding a dynamic type will allow the user to pass multiple arguments to their script if they require extra logic.

I think `useLocatorStrategy` is a good name and its intuitive since the user will attach strategies via `browser.addLocatorStrategy`.

**Comment by christian-bromann** [comment]
> on second thought, having both `custom$` and `custom$$` might be confusing since you have to add the strategies before you use them.

In what way is this confusing?

> It would be better to consider renaming the command to something more explicit like `useLocatorStrategy`

My concern is that it is not aligned with other command names such as `react$/$$` or `shadow$/$$`

> and we always return one element since it depends on the script passed by the user.

This is not ensured by the user script. If for whatever reason the script fetches multiple elements, multiple elements will be returned.

**Comment by baruchvlz** [comment]
There is no real difference between `custom$/$$` and `useLocatorStrategy` and you have a good point on keeping the methods consistent. I'll add the methods to both browser and element objects.

File: packages/webdriverio/src/commands/browser/custom$$.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        throw Error('No strategy found for ' + strategyName)
+    }
+
+    let res = await this.execute(strategy, strategyArgument)
+
+    if (!Array.isArray(res)) {
+        res = [res]
+    }
+
+    return await getElements.call(this, strategy, res)
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/browser/custom$$.js]:
```suggestion
    return await getElements.call(this, strategy, res.filter((el) => Boolean(el) && typeof el[ELEMENT_ID] === 'string'))
```

Let's add a check to ensure that the returned object has an actual element reference.

File: packages/webdriverio/src/commands/browser/custom$.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+async function custom$ (strategyName, strategyArguments) {
+    const strategy = this.strategies.get(strategyName)
+
+    if (!strategy) {
+        throw Error('No strategy found for ' + strategyName)
+    }
+
+    let res = await this.execute(strategy, strategyArguments)
+
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/browser/custom$.js]: same here

File: packages/webdriverio/src/commands/browser/custom$$.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+export default runFnInFiberContext(custom$$)
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/browser/custom$$.js]: This is not necessary. We don't do that with any command.

File: packages/webdriverio/src/commands/element/custom$$.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * @alias custom$$
+ * @param {String} strategyName
+ * @param {Any} strategyArguments
+ * @return {Element}
+ */
+import { runFnInFiberContext } from '@wdio/utils/build/shim'
+import { getElements } from '../../utils/getElementObject'
+import { getBrowserObject } from '../../utils'
+
+async function custom$$ (strategyName, strategyArgument) {
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/element/custom$$.js]: since it is the same implementation you can just require it from the browser directory, no?

Comment (baruchvlz) [reply packages/webdriverio/src/commands/element/custom$$.js]: @christian-bromann It's missing the `parent` element as in the `element.custom$` function. Will fix.

File: packages/webdriverio/src/commands/browser/custom$$.js (hunk 3 above)
Comment (baruchvlz) [reply packages/webdriverio/src/commands/browser/custom$$.js]: @christian-bromann I wasn't able to run this in sync mode without the fiber context wrapper. Will look for another way.

File: packages/webdriverio/src/commands/browser/custom$$.js (hunk 1 above)
Comment (baruchvlz) [reply packages/webdriverio/src/commands/browser/custom$$.js]: @christian-bromann  Not entirely sure what the goal of this is. Wouldn't getElements error if the res doesn't contain a WDIO element? If not, should `custom$/$$` throw the error?

File: packages/webdriverio/src/commands/browser/custom$$.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
+/**
+ *
+ * The `customs$` allows you to use a custom strategy declared by using `browser.addLocatorStrategy`
+ *
+ * <example>
+    :pause.js
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/browser/custom$$.js]:
```suggestion
    :example.js
```

File: packages/webdriverio/src/commands/browser/custom$$.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    it('should get all the plugin wrapper buttons', () => {
+        browser.url('https://webdriver.io');
+        browser.addLocatorStrategy('myStrat', (selector) => {
+            return document.querySelectorAll(selector)
+        })
+
+        const pluginWrapper = browser.custom$$('myStrat', '.pluginWrapper')
+
+        console.log(pluginWrapper.length) // 4
+    });
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/browser/custom$$.js]:
```suggestion
    })
```

File: packages/webdriverio/src/commands/browser/custom$$.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
+/**
+ *
+ * The `customs$` allows you to use a custom strategy declared by using `browser.addLocatorStrategy`
+ *
+ * <example>
+    :pause.js
+    it('should get all the plugin wrapper buttons', () => {
+        browser.url('https://webdriver.io');
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/browser/custom$$.js]:
```suggestion
        browser.url('https://webdriver.io')
```

File: packages/webdriverio/src/commands/browser/custom$.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
+/**
+ *
+ * The `custom$` allows you to use a custom strategy declared by using `browser.addLocatorStrategy`
+ *
+ * <example>
+    :pause.js
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/browser/custom$.js]:
```suggestion
    :example.js
```

File: packages/webdriverio/src/commands/browser/custom$.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    it('should fetch the project title', () => {
+        browser.url('https://webdriver.io');
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/browser/custom$.js]:
```suggestion
        browser.url('https://webdriver.io')
```

File: packages/webdriverio/src/commands/element/custom$$.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
+/**
+ *
+ * The `customs$` allows you to use a custom strategy declared by using `browser.addLocatorStrategy`
+ *
+ * <example>
+    :pause.js
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/element/custom$$.js]:
```suggestion
    :example.js
```

File: packages/webdriverio/src/commands/element/custom$$.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    it('should get all the plugin wrapper buttons', () => {
+        browser.url('https://webdriver.io');
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/element/custom$$.js]:
```suggestion
        browser.url('https://webdriver.io')
```

File: packages/webdriverio/src/commands/element/custom$$.js [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        browser.addLocatorStrategy('myStrat', (selector) => {
+            return document.querySelectorAll(selector)
+        })
+
+        const pluginRowBlock = browser.custom$('myStrat', '.pluginRowBlock')
+        const pluginWrapper = pluginRowBlock.custom$$('myStrat', '.pluginWrapper')
+
+        console.log(pluginWrapper.length) // 4
+    });
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/element/custom$$.js]:
```suggestion
    })
```

File: packages/webdriverio/src/index.js [hunk 13]
```diff
@@ -61,6 +61,14 @@ export const remote = async function (params = {}, remoteModifier) {
         )
     }
 
+    instance.addLocatorStrategy = (name, script) => {
```
Comment (christian-bromann) [inline packages/webdriverio/src/index.js]: How about we move this function into a utility as follows:
```js
instance.addLocatorStrategy = addLocatorStrategyHandler(instance)
```

the utility then looks like

```js
export function addLocatorStrategyHandler (scope) {
    if (scope.strategies.get(name)) {
        throw new Error(`Strategy ${name} already exists`)
    }

    scope.strategies.set(name, script)
}
```

Comment (baruchvlz) [reply packages/webdriverio/src/index.js]: Wouldn't it be more like this?

```js
export function addLocatorStrategyHandler (scope) {
    return (name, script) => {
        if (scope.strategies.get(name)) {
            throw new Error(`Strategy ${name} already exists`)
        }

        scope.strategies.set(name, script)
    }
}
```

Comment (christian-bromann) [reply packages/webdriverio/src/index.js]: yes

**Comment by christian-bromann** [comment]
@baruchvlz can you rebase?

**Pull request merged**